### PR TITLE
HPCC-14115 LDAP Enumerations need to support paging

### PR DESCRIPTION
**Pull request by RussWhitehead**
LDAP imposes a limit of 1,000 on the number of entries that can be returned
from any search. Once there are over 1,000 matches to any search, an error
is returned and you cannot even get the first 1000.  This PR implements a
CPagedLDAPSearch class, which wraps code to perform an LDAP Paged Search,
and uses it in any ILDAPClient calls that could return this many entries.
It also implements a countEntries method, which is used in several places
within LDAP clients.  This includes:

retrieveUsers
getResources
getResourcesEx
getAllGroups
getGroupMembers
countEntries

Signed-off-by: Russ Whitehead william.whitehead@lexisnexis.com

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

**Comment by RussWhitehead** [comment]
Tested for memory leaks with Valgrind, no new ones introduced

**Comment by RussWhitehead** [comment]
Tested successfully with Microsoft Active Directory and 389 Directory Server

**Comment by wangkx** [comment]
@RussWhitehead back to you.

**Comment by RussWhitehead** [comment]
@wangkx Back at you

**Comment by wangkx** [comment]
@RussWhitehead looks good.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge and close

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -981,6 +981,202 @@ class CLDAPMessage
...
+
+            if (m_pCookie)
+            {
+                ber_bvfree(m_pCookie);
+                m_pCookie = NULL;
+            }
+
+            if (rc != LDAP_SUCCESS)
+            {
+                ldap_control_free(pageControl);
```
Comment (richardkchapman) [inline system/security/LdapSecurity/ldapconnection.cpp]: I would have been inclined to create a little wrapper class to contain the pageControl ptr and ensure it was freed on destruction - that way you don't need to check every possible return/throw to make sure it is freed...

But I'm not going to hold up this merge for that suggestion... maybe next time.

**Pull request merged**